Speeding up GPU clustering using smarter download strategy and memory allocations

FabianSchuetze opened the pull request:
I stumbled upon the same issue mentioned in #2703 while familiarizing myself with the GPU-related codebase of PCL. A CPU flamegraph showed  about 1/3 of the runtime is spent on resizing a vector: 
![perf_orig](https://user-images.githubusercontent.com/10357496/112655136-63556200-8e50-11eb-8479-a832527e09d9.jpg)
The resize function is located in the [device_array.hpp](https://github.com/PointCloudLibrary/pcl/blob/97cb436308e240c34f49559ba3b89b6cdfc2e5ca/gpu/containers/include/pcl/gpu/containers/impl/device_array.hpp#L79) codebase. The arrays are constructed in the loop (line 100) of [gpu_extract_custer.hpp](https://github.com/PointCloudLibrary/pcl/blob/97cb436308e240c34f49559ba3b89b6cdfc2e5ca/gpu/segmentation/include/pcl/gpu/segmentation/impl/gpu_extract_clusters.hpp#L100). Pre-allocating the array to the maximum possible size can mostly eliminate the memory allocations as documented by the flamegraph with the revised code:
![perf](https://user-images.githubusercontent.com/10357496/112656122-5d13b580-8e51-11eb-9b96-dafba606bb47.jpg)  I tried to optimize the Cuda memcopies that also incur significant times by using pinned host memory, but that did not lead to a noticeable improvement.  The GPU code is still significantly slower than the CPU version, though. I believe this is due to the sequential nature of the program, and data copies between the host and device memory.

I would be happy to work more on this issue if further PRs are welcome in this field. Maybe somebody also has an idea for improving GPU-based segmentation or other ideas to work on the GPU-related codebase of PCL! 

P.S. The function [pcl::gpu::extractEuclideanClusters](https://github.com/PointCloudLibrary/pcl/blob/97cb436308e240c34f49559ba3b89b6cdfc2e5ca/gpu/segmentation/include/pcl/gpu/segmentation/impl/gpu_extract_clusters.hpp#L45) is verbose by default, should we maybe also change this?

mvieth commented:
Nice idea, and the flame graphs look promising
- Could you do some simple benchmarking to find out how much faster than the current implementation your proposal is?
- What do you think about moving the creation of the vectors out of the for-loop (moving it to line ~70)? Could that further reduce the number of allocations?
- I am not sure the value you use for `reserve` is the correct one. I think `data` is `queries_host.size() * max_answers` large, so up to `host_cloud_->size() * max_answers`. But that can be very large (several gigabytes), and I wonder how much difference the `reserve` makes anyway? Some simple benchmarks here would be great as well
- Please also move the comment "Host buffer for results"
- Yes, if you like, you can also change the logging behaviour, e.g. with the PCL_DEBUG macro
- If you like to further work on the GPU clustering, you could check if the value "10" (line 103) is the best option or if other options make the clustering faster. You could also make the value settable as the `@todo` suggests

FabianSchuetze commented:
Thank you for the kind and detailed feedback, @mvieth - that was very helpful to me! I will address the comments and update the PR.

FabianSchuetze commented:
Thank you again for your detailed comments!  I tried to address these comments as follows:

- With a pointcloud from my living room containing 198,835 points, the computation time drops from 9.40 to  4.23s. Similarly, for the pointcloud "five_people.pcd" from PCD's test folder (comprising 307,200 points),  the duration falls from 10.63s to 7.09 seconds.  
- Moving the vectors one scope up and not calling reserve was an excellent good suggestion - thank you! With the persistent vector, calling resizing leads to few memory allocations. 
- I did move the comment "Host buffer for results" and moved the logging into the PCL_DEBUG macro. 
- I also experimented with changing the threshold of offloading the computation to the GPU and compared values of (10, 50, 100, 1000) This is a losing battle for the GPU at the moment: Increasing the threshold reduces total computation time. As documented by the flamegraph, the latency for copying data between host and device biases comparison. I would increase the threshold to 100, but I am not sure if the comparison is entirely fair. 

I hope this address all the comments you made and that I did not forget anything. Is there anything else you suggest looking at? I also thought more about the problem and have two other ideas for improving it: 

According to the flamegraph, 9 percent of the time is spent in malloc inside the [create function](https://github.com/PointCloudLibrary/pcl/blob/78d1f444f3be24332114694da6d7b54336687f0e/gpu/containers/src/device_memory.cpp#L125) of the `DeviceArray`. The `create` function destroys (if needed) memory and fully allocates new memory again. We have to call the create function because the device array is always local to the scope and because the device array does not have a resize function. Such a resize function, for example, is also part of [thrust](https://docs.nvidia.com/cuda/thrust/index.html). Do you think it would be interesting for us to expand the API of the `DeviceArray` to permit resizing the array? 

Apart from resizing the array, copying data from the device to the host requires the most time (about 2/3 in the current form). Looking at lines [149ff](https://github.com/PointCloudLibrary/pcl/blob/78d1f444f3be24332114694da6d7b54336687f0e/gpu/segmentation/include/pcl/gpu/segmentation/impl/gpu_extract_clusters.hpp#L149), I think we could avoid copying the data as the only relevant computation is obtaining the indices from `data[qp_r + qp * max_answers]`. I guess that we could do this all on the GPU threads. 

Do you think any of these ideas are worthwhile to pursue? Or would you alternatively suggest working on other areas of the GPU section?

FabianSchuetze commented:
I'm sorry to see the build partially failing. Could somebody kindly help me understanding why that happened? I looked through the logs and found only one statement that might report an error:
```
2021-03-29T12:42:13.4574189Z   LINK : the 32-bit linker (C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\VC\Tools\MSVC\14.16.27023\bin\HostX86\x86\link.exe) ran out of heap space and is going to restart linking with a 64-bit linker
2021-03-29T12:42:13.5194291Z   LINK : restarting link with 64-bit linker `C:\Program Files\Git\usr\bin\link.exe'
2021-03-29T12:42:18.9432564Z   /c/Program Files (x86)/Microsoft Visual Studio/2017/Enterprise/VC/Tools/MSVC/14.16.27023/bin/HostX86/x86/link: cannot create link 'Â â– /' to '/ERRORREPORT:QUEUE': No such file or directory
```
However, the build continues for a while afterwards and I suspect I misinterpret the error. I am grateful for any hints or suggestions to understand the build results!

mvieth commented:
So a speed-up of 2.2 and 1.5 respectively - that is really nice.
Regarding the offloading threshold: can you give an estimate of how much 10 vs 100 changes? I made some quick tests and it seems to barely change the time.
Regarding adding a resize function to `DeviceArray`: you can try that out if you like, but I don't know how much impact it would make if it only accounts for 9% of the time spent.
Regarding copying the data from device to host: I am not sure how you would completely avoid copying the data - by calling `ptr()` and accessing the data from there? Not sure if that works/improves the performance, but might be worth a try. Another idea to reduce copying: there are many elements in `data` that are copied, but never accessed, since there are blocks of size `max_answers` that might not be filled completely. If we would first download `sizes`, and then somehow only download those chunks of `data` that we actually need, that might improve performance.
@larshg If you like, have a look at our discussion. Since you worked on the GPU clustering before, you might have a good suggestion
P.S.: Regarding the failed test: the Windows CI sometimes runs out of memory - I restarted it and the check should probably pass now.

FabianSchuetze commented:
Thank you very much, @mvieth, for your detailed and thoughtful comments - they again helped me understand the problem much better! 

I'm surprised you did not experience a significant change in run-time.  With my configuration, the run-time changed almost monotonically. I will report detailed benchmarks again. 

Your comment that we download more data than necessary was striking to me, I did not realize this first. For a sampled iteration of the nested for loop the selected indices `qp_r + qp * max_answers` are: `0, 1, ...,20, 25000, 25001, ..., 25024, 50000, ...`. Although only a small fraction of the indices of `data` are accessed, we donwload the entire array. We could indeed try to filter the array on the device before downloading it. That is a fascinating challenge! 

P.S. Thanks for running the build again - its reassuring  to see that not only my computer runs out of memory when building PCL when using all possible cores.

larshg commented on gpu/segmentation/include/pcl/gpu/segmentation/impl/gpu_extract_clusters.hpp:
```diff
@@ -60,14 +61,17 @@ pcl::gpu::extractEuclideanClusters (const pcl::PointCloud<pcl::PointXYZ>::Ptr  &
 
+  // Host buffer for results
+  std::vector<int> sizes, data;
+
```
Was the intention to remove the preallocations again?

FabianSchuetze replied:
We were not sure about the correct size and realized that virtually all the performance gains can be realized by just keeping the vectors persistently in scope and let the resize function find the correct capacity.

larshg commented:
I'm just testing it out as well.
However after the offloading threshold was set to 100, I didn't get any clusters. Seems like the CPU version is faulty here:
https://github.com/PointCloudLibrary/pcl/blob/97cb436308e240c34f49559ba3b89b6cdfc2e5ca/gpu/segmentation/include/pcl/gpu/segmentation/impl/gpu_extract_clusters.hpp#L105-L110

For each query points, the radius search is performed, but the results are not accumulated and hence, it never finds the required sizes of clusters.

Also, it seems, with my preliminary tests of setting it to 200 (No GPU activity with the pointcloud / parameters I use), the "CPU implementation of the GPU algorithm" is faster than the euclidian clustering in pcl_segmentation (using a K-D tree).

Some more investigation is required.

How was #3860 progressing @mvieth  :) ?

FabianSchuetze commented:
Thanks for the detailed and very helpful comments, @larshg!  Your comments about the difference between the CPU and GPU version are indeed intriguing and I am happy to investigate it more thoroughly!

kunaltyagi commented on gpu/segmentation/include/pcl/gpu/segmentation/impl/gpu_extract_clusters.hpp:
```diff
@@ -132,7 +133,8 @@ pcl::gpu::extractEuclideanClusters (const pcl::PointCloud<pcl::PointXYZ>::Ptr  &
       {
-        std::cout << " GPU: ";
+        PCL_DEBUG(" GPU: ");
+        std::cout << "GPU " << std::endl;
```
Remove this in favor of the debug?

kunaltyagi commented on gpu/segmentation/include/pcl/gpu/segmentation/impl/gpu_extract_clusters.hpp:
```diff
@@ -96,13 +100,10 @@ pcl::gpu::extractEuclideanClusters (const pcl::PointCloud<pcl::PointXYZ>::Ptr  &
     do
     {
-      // Host buffer for results
-      std::vector<int> sizes, data;
```
Should we be calling `{data,size}.clear()` somewhere? Or is the intention to persist their data (and memory) between the do-while loops?

FabianSchuetze replied:
I think calling clear on the vector is a good idea as it shows the programmers intent clearly! I have now called it at the beginning of the for loop.

FabianSchuetze commented:
@larshg Thank you again for pointing out that the obtained clusters are a function of the GPU offloading threshold! I did verify this, which left me puzzled for a while. I adjusted the scope of the loop you pointed out, but that didn't help. I will try to figure out what is going on and then write again. 

@mvieth  I refrained from benchmarking the run-time of the algorithm as a function of the offloading threshold. I will first try to understand why changing the threshold causes the results to differ and then measure run-time performances. 

@kunaltyagi  Thanks for the comments! I did adjust the code in response to your observations, and I think it improves the code. 

At all three of you: Thank you so much for taking a look at this PR and for guiding me through the process - this is indeed a pleasure!

FabianSchuetze commented:
First: Thank you all again for the excellent feedback so far! It is really a pleasure to work on this. 

I continued working on the issue and made progress on the program's efficiency. As @mvieth  suggested, we currently download potentially too much data from the device to host. A more judicious download gives us speed gains of a factor of 3. I updated the PR to document the ideas in code. However, I do not think this code is pretty, and I am not sure if it's a good idea to merge it. To keep the discussion focused, I opened another draft PR to expand the API of the DeviceArray to allow users to download data more effectively. 

Here are some performance benchmarks (taken for the pcd test data)
Dataset: (turning the cpu version of the GPU part of; i.e. if (queries_host.size() <= 0) go to CPU; else GPU)

Dataset | rops_cloud (804k) | bunny (11K) | cturtle (771K)
--|--|--|--|
CPU| 1.31 | 0.01 | 3.18|
GPU| 0.49 | 0.08 | 1.63|
Master branch| 1.61| 0.07 | 8.63 | 

The flamegraphs reflect the timing updates too. The graph still highlights further possibilities for improvement of the GPU version. However, the CPU code dominates the timing for the first time :-).  
![perf](https://user-images.githubusercontent.com/10357496/113475210-6f56aa80-9474-11eb-99a3-28087dfd4bd8.png)


I realize that the problem of different results depending on the threshold parameter persists! Thanks again for highlighting this, @larshg! Although I am thinking about it, I am still tapping a bit in the dark but will keep trying. What do you think about the changes and what are open questions for you? 

P.S. I cannot use a few of the test datasets because FLANN fails on the CPU. These are: table_scene_mug_stereo_textured,  office1d, or five_people. The (abbreviated) error is "Invalid (NaN, Inf) point coordinates given to radiusSearch!"' failed. I have looked for an issue but didn't see one. Does somebody know an issue?

FabianSchuetze commented:
Ha! I think things are falling in place! 

Thank you again, @larshg ,  for identifying that the number of clusters changes as the offloading parameter changes. As you speculated, not all found indices found inside the method `radiusSearchHost` got propagated through the program. The search method clears the vector of previously found indices, and thus these indices got lost. I now obtain the same results on the dataset `five_people` for a broad range of offloading parameters.  @larshg , can you kindly check the new program on the dataset on which you identified the problem at first? That would be wonderful!

larshg commented:
Yes, thats pretty much the same thing I thought would be required. Not sure if you can use a `move_inserter` to avoid making a copy?

And maybe use `pcl::indices`(its currently defined as the same, but makes it easier to change type of indices) instead of std::vector if searchHost returns indices?

kunaltyagi commented on gpu/segmentation/include/pcl/gpu/segmentation/impl/gpu_extract_clusters.hpp:
```diff
@@ -41,6 +41,25 @@
+#include <cuda_runtime_api.h>
+#include <cuda.h>
+std::vector<int> economical_download(const pcl::gpu::NeighborIndices& indices,
+        std::vector<int>& sizes, int max_answers) {
```
const reference?

FabianSchuetze replied:
We do modify `sizes`, by downloading data into it `indices.sizes.download(sizes);`, We could download the data outside of the function, and then pass a const reference. Would you think that is a good idea?

kunaltyagi replied:
I don't think moving the download outside the function will affect either the name of the intended usage, so either way (const ref + download outside, or mut ref + comment for the reason) is fine by me

FabianSchuetze replied:
Sounds great to me - thank you!

FabianSchuetze commented on gpu/segmentation/include/pcl/gpu/segmentation/impl/gpu_extract_clusters.hpp:
```diff
@@ -96,24 +121,23 @@ pcl::gpu::extractEuclideanClusters (const pcl::PointCloud<pcl::PointXYZ>::Ptr  &
         {
           // Execute the radiusSearch on the host
-          tree->radiusSearchHost(queries_host[p], tolerance, data, max_answers);
+          std::vector<int> tmp;
```
Great idea - thanks!

FabianSchuetze commented on gpu/segmentation/include/pcl/gpu/segmentation/impl/gpu_extract_clusters.hpp:
```diff
@@ -41,6 +41,25 @@
+    for (std::size_t qp = 0; qp < sizes.size(); qp++) {
+        int begin = qp * max_answers;
+        const int* pdata = indices.data.ptr() + begin;
+        std::vector<int> tmp(sizes[qp]);
```
What do you want to move out of the loop? I think I can move the initialization of the vector out of the loop. I am not so sure about the other two lines.

kunaltyagi replied:
The vector can be moved outside the loop, with the first line of the loop clearing the vector. This will reduce the allocations from the current 1 per iteration.

If needed, we can find the max size and reserve the vector causing exactly 1 allocation

FabianSchuetze replied:
Great - then I understood what you mean. Thanks for the explanation. I have moved the construction of the vector outside and the vector is also cleared each time.

FabianSchuetze commented:
Thank you, @kunaltyagi,  for reviewing the PR - that was very helpful indeed! I updated the PR accordingly. 

Thank you also very much, @larshg! I replaced all instances of `std::vector<int>` with `pcl::indices`. I think this is a better solution - thank you! Moving a`vector<int>` induces a copy because the values from one contiguous memory location have to be placed in the other contiguous memory location. If we had a vector of strings (or any other type that manages a pointer to a memory location), move inserters would be perfect. 

What are your opinions about placing the cuda interaction into the `DeviceArray` class? I think the following lines of code could be wrapped inside a `download` function of `DeviceArray`: 
```c++
#include <cuda_runtime_api.h>
#include <cuda.h>
...
const std::size_t bytes = (sizes[qp]) * sizeof(int);
cudaMemcpy(&tmp[0], pdata, bytes, cudaMemcpyDeviceToHost);
cudaDeviceSynchronize();
```
I think all these lines could be placed inside a function overloading the `download` member function of the `DeviceArray`. I tried to describe it in a ticket #4689 .

kunaltyagi commented on gpu/segmentation/include/pcl/gpu/segmentation/impl/gpu_extract_clusters.hpp:
```diff
@@ -41,6 +41,26 @@
+    indices.sizes.download(sizes);
+    pcl::Indices results, tmp;
+    for (std::size_t qp = 0; qp < sizes.size(); qp++) {
+        const int begin = qp * max_answers;
```
`std:::size_t` instead?

FabianSchuetze replied:
Excellent!

kunaltyagi commented on gpu/segmentation/include/pcl/gpu/segmentation/impl/gpu_extract_clusters.hpp:
```diff
@@ -41,6 +41,26 @@
+pcl::Indices economical_download(const pcl::gpu::NeighborIndices& indices,
+        pcl::Indices& sizes, int max_answers) {
+    indices.sizes.download(sizes);
+    pcl::Indices results, tmp;
```
`tmp` HAS to be of type `std::vector<int>` due to L55

FabianSchuetze replied:
That makes sense - thank you!

kunaltyagi commented on gpu/segmentation/include/pcl/gpu/segmentation/impl/gpu_extract_clusters.hpp:
```diff
@@ -41,6 +41,26 @@
+#include <cuda.h>
+
+pcl::Indices economical_download(const pcl::gpu::NeighborIndices& indices,
+        pcl::Indices& sizes, int max_answers) {
```
As commented before, please add a comment for the mut ref, or use a const ref by moving the download outside the function

The order of the arguments could be improved. A possible alternative:
```cpp
namespace pcl {
namespace detail {
// There's no point in performing an allocation. Just out parameter for returning the results
// All these vectors can be further moved out of the big for loop in L96
void economical_download(source_indices, points_per_cluster, result, indices_size, indices_buffer) {
    // do the thing
}
}  // ns details
}  // ns pcl
```

FabianSchuetze replied:
I very much agree that the order of the args could be better - thank you! Thank you also for suggesting to pack the function in a namespace. I did all that. Does the function now look better to you?

FabianSchuetze commented:
Thank you, @kunaltyagi , for your comments - they were very helpful for me!

kunaltyagi commented on gpu/segmentation/include/pcl/gpu/segmentation/impl/gpu_extract_clusters.hpp:
```diff
@@ -60,17 +86,22 @@ pcl::gpu::extractEuclideanClusters (const pcl::PointCloud<pcl::PointXYZ>::Ptr  &
   queries_device_buffer.create(max_answers);
 
+  // Host buffer for results
+  pcl::Indices sizes, data, cpu_tmp, gpu_indices;
```
Similar to cpu_tmp, we can create a `gpu_tmp` here to replace the indices created in the new function on L54. That'll reduce more allocations for the GPU case (due to the loop at L101)

FabianSchuetze replied:
As I outline below, I would suggest using either lambda and gpu_tmp or continue using the function and use gpu_tmp as a local variable.

kunaltyagi commented on gpu/segmentation/include/pcl/gpu/segmentation/impl/gpu_extract_clusters.hpp:
```diff
@@ -132,35 +162,34 @@ pcl::gpu::extractEuclideanClusters (const pcl::PointCloud<pcl::PointXYZ>::Ptr  &
-        result_device.data.download (data);
+        result_device.sizes.download(sizes);
+        gpu_indices.clear();
+        pcl::detail::economical_download(result_device, sizes, max_answers, gpu_indices);
```
I'm wondering if a function is the right fit, instead of a lambda with capture. Thoughts?

Pros:
* capture the variables properly, so the call syntax is less noisy
    ```
    economical_download(result_device, max_answers, gpu_indices);  // with lambda that captures other variables
    economical_download(result_device, sizes, max_answers, gpu_indices, gpu_tmp);  // with function
* No additional symbol added under detail namespace

Cons:
* lambda will be placed around L100, and I've found others complaining about readability differences between free functions and non-immediately used lambdas
* lambda might have to be mutable (for `gpu_tmp`)

FabianSchuetze replied:
Thank you so much for the comment, @kunaltyagi! I usually appreciate the lambda's conciseness and happily thought about its use here. I modified the program as follows:

```c++
  pcl::Indices sizes, data, cpu_tmp, gpu_tmp, gpu_indices;
  auto economical_download = [&](const auto& source_indices,
        const auto& buffer_indices, auto& downloaded_indices) {
    for (std::size_t qp = 0; qp < buffer_indices.size(); qp++) {
        std::size_t begin = qp * max_answers;
        const int* const pdata = source_indices.data.ptr() + begin;
        gpu_tmp.resize(buffer_indices[qp]);
        const std::size_t bytes = (buffer_indices[qp]) * sizeof(int);
        cudaMemcpy(&gpu_tmp[0], pdata, bytes, cudaMemcpyDeviceToHost);
        cudaDeviceSynchronize();
        downloaded_indices.insert(downloaded_indices.end(), gpu_tmp.begin(), gpu_tmp.end());
    }
  ;};
```
The function now has one fewer argument (`max_answers`), and `gpu_tmp` gets captured. I very much appreciate these two aspects. Did you have another idea for writing the lambda?

Concluding, I am undecided on whether introducing the lambda is beneficial. The lambda is a bit long and interrupts the code's flow for my taste. I would suggest using either lambda and `gpu_tmp` or continue using the function and use `gpu_tmp` as a local variable. The allocations do not have a substantial impact, and the vector gets used only locally. Do you have a preference?

kunaltyagi replied:
Yeah, that would have been my lambda as well.

> The allocations do not have a substantial impact, and the vector gets used only locally.

This is curious. Assuming buffer_indices are not all same, the resize would incur a cost right before a CPU write. Perhaps my intuition is wrong.

The current speed-up is anyways quite good :)

FabianSchuetze replied:
Great to see we would have written identical lambdas! And yes, oh boy, these allocations are playing tricks with me all day long! 

However, the [most recent flamegraph](https://user-images.githubusercontent.com/10357496/113475210-6f56aa80-9474-11eb-99a3-28087dfd4bd8.png) documents that allocations are not substantial. The right end of the image shows a yellow vertical bar named `pcl::gpu::EuclidianClust`.  The bar's length is proportional to the time spent inside this function, and vertical bars stacked above denote the function's callees. As shown, `economical_download` is predominantly waiting for Cuda-related calls to complete. The small vertical slice whose name is too small to be read on this image contains code for the vector memory allocations. In contrast, the [first flamegraph](https://user-images.githubusercontent.com/10357496/112655136-63556200-8e50-11eb-8479-a832527e09d9.jpg) highlights that the unoptimized program waited a long time for `std::vector::resize` to complete.

kunaltyagi replied:
Then no point in optimizing this :)

kunaltyagi reviewed (commented):
Just some thoughts about 1 final memory allocation

kunaltyagi commented on gpu/segmentation/include/pcl/gpu/segmentation/impl/gpu_extract_clusters.hpp:
```diff
@@ -132,35 +162,34 @@ pcl::gpu::extractEuclideanClusters (const pcl::PointCloud<pcl::PointXYZ>::Ptr  &
-            queries_host.push_back ((*host_cloud_)[data[qp_r + qp * max_answers]]);
+        for (int idx : gpu_indices) {
+            if (processed[idx])
+                continue;
+            processed[idx] = true;
+            queries_host.push_back((*host_cloud_)[idx]);
             found_points++;
-            r.indices.push_back(data[qp_r + qp * max_answers]);
-          }
+            r.indices.push_back(idx);
         }
```
2 thoughts:
* this is exactly the same as L120-130, thanks to economical download
* L189 will give wrong result since data is not used in this branch of if-else, instead gpu_indices are used

FabianSchuetze replied:
Ha! You pointed out an intriguing symmetry (if I understood what you meant). We can indeed refactor the code nicely. I made an attempt to do that. What do you think about it?

FabianSchuetze commented on gpu/segmentation/include/pcl/gpu/segmentation/impl/gpu_extract_clusters.hpp:
```diff
@@ -41,6 +41,31 @@
+    }
+}
+} // ns detail
+} // ns pcl
```
Indeed - fixed!

kunaltyagi commented on gpu/segmentation/include/pcl/gpu/segmentation/impl/gpu_extract_clusters.hpp:
```diff
@@ -96,71 +124,57 @@ pcl::gpu::extractEuclideanClusters (const pcl::PointCloud<pcl::PointXYZ>::Ptr  &
+        continue;
+
+      // Process the results
+      for(int idx : data)
```
`auto` would fit here quite well

FabianSchuetze replied:
ok

kunaltyagi commented on gpu/segmentation/include/pcl/gpu/segmentation/impl/gpu_extract_clusters.hpp:
```diff
@@ -96,71 +124,57 @@ pcl::gpu::extractEuclideanClusters (const pcl::PointCloud<pcl::PointXYZ>::Ptr  &
-            queries_host.push_back ((*host_cloud_)[data[qp_r + qp * max_answers]]);
-            found_points++;
-            r.indices.push_back(data[qp_r + qp * max_answers]);
-          }
-        }
+        if(processed[idx])
+          continue;
+        processed[idx] = true;
+        queries_host.push_back ((*host_cloud_)[idx]);
+        found_points++;
+        r.indices.push_back(idx);
       }
```
The braces need formatting

FabianSchuetze replied:
Oh, sorry, I tried to fix it! Is it correct now? I have a question though: When I clang-format the file `gpu_extract_clusters.hpp` many lines of the file change. Am I doing something wrong or does the same maybe also happen to you too?

kunaltyagi replied:
`gpu` module isn't formatted yet. So your experience is correct. [The plugin I use](https://github.com/rhysd/vim-clang-format) allows me to format only a selection of lines instead of all the lines in the file.

You can either:
* use similar settings for some plugin in your fav editor
* create an additional formatting commit (at the start or end) so we can skip that/select that easily during code-review

FabianSchuetze replied:
Ha - thanks for pointing out this feature in vim clang-format! I always only format the entire document, never a few lines. Shall I maybe make a batch PR clang-formatting the GPU codebase?

kunaltyagi commented on gpu/segmentation/include/pcl/gpu/segmentation/impl/gpu_extract_clusters.hpp:
```diff
@@ -41,6 +41,31 @@
+namespace detail {
+
+// Downloads only the neccssary cluster indices from the device to the host.
+void economical_download(const pcl::gpu::NeighborIndices& source_indices,
```
Shall we make it inline (or move the impl to cpp or make it templated)? It is a non-template function and I don't feel comfortable letting it go inside a header (multiple declarations are a possibility by inclusion in 2 cpp files which are linked together)

FabianSchuetze replied:
This is an excellent question! While I understand your concerns, I am not used to working on such large codebases and have no experience in making an informed judgment. Do you have a recommendation?

kunaltyagi replied:
@larshg thoughts?

larshg replied:
Based on a quick skimming of https://stackoverflow.com/questions/1932311/when-to-use-inline-function-and-when-not-to-use-it

I would go for moving to a .cpp file.

- Its not a small function
- Its I/O bound (GPU?)

FabianSchuetze replied:
Thank you very much, @larshg  for your comments! Do you know in which cpp file you would place the function? I see that [`extract_clusters.cpp`](https://github.com/PointCloudLibrary/pcl/blob/master/gpu/segmentation/src/extract_clusters.cpp) exists, but this file does not include [`gpu_extract_clusters.h`](https://github.com/PointCloudLibrary/pcl/blob/master/gpu/segmentation/include/pcl/gpu/segmentation/gpu_extract_clusters.h) yet and has no implementations in it.  

The function `economical_download` is indeed IO bound, spending almost all its time on device-host interaction.

larshg replied:
Just include gpu_extract_clusters.h and add the implementation. Seems to be have done something similar here: https://github.com/PointCloudLibrary/pcl/blob/master/recognition/src/cg/hough_3d.cpp

FabianSchuetze replied:
great - thanks @larshg! I will do exactly that.

kunaltyagi replied:
PS: The impl hpp file should be including the h file for correct behavior, and including the h file in cpp file wouldn't be need if that's the case

FabianSchuetze replied:
Thank you both, @larshg and @kunaltyagi,  for your detailed and helpful instructions! I placed the function `economical_download` in a source file and declared it in the header file. The function is a static member as I thought this is the most fitting namespace. Is this in line with your ideas?

kunaltyagi reviewed (commented):
LGTM otherwise :)

Thanks @FabianSchuetze for bearing with us

FabianSchuetze commented:
> LGTM otherwise :)
> 
> Thanks @FabianSchuetze for bearing with us

It was a pleasure to work on it - thank you so much for your support @kunaltyagi ! 

I have another question as this issue comes to an end:  I would like to continue working on the GPU code. Issues #4443 or #2218 seem interesting to me. Alternatively, the GPU code seems to lack a correspondence estimation, and I would be happy to work on this. I would be thrilled to see a GPU version of the ICP algorithm and think this should be the next step. Do you have an idea of which feature/issue to prioritize?

kunaltyagi commented on gpu/segmentation/include/pcl/gpu/segmentation/gpu_extract_clusters.h:
```diff
@@ -132,6 +132,17 @@ namespace pcl
+         * param buffer_indices indices denoting positions to download
+         * param buffer_size length of a memory buffer
+         * param downloaded_indices memory buffer of the downloaded indices */
+        static void
```
I don't think this should be:
* public (strongly)
* class member (meh, it can be but does it have to? It'll become part of the API and ABI and add maintenance overhead)

FabianSchuetze replied:
Thank you for your detailed and very instructive comment, @kunaltyagi ! I put the function inside the class only with great trepidation but thought this was the best fit. However, your comments point out well why this was not a good choice. I placed the function thus again inside the `pcl::detail` namespace. Does that address your concerns?

kunaltyagi commented on gpu/segmentation/include/pcl/gpu/segmentation/gpu_extract_clusters.h:
```diff
@@ -48,6 +48,18 @@
+ * param buffer_size length of a memory buffer
+ * param downloaded_indices memory buffer of the downloaded indices */
+void
+economical_download(const pcl::gpu::NeighborIndices& source_indices,
```
We can move this to the hpp file

FabianSchuetze replied:
Yep, I did that - thanks for your comment! :-)

kunaltyagi reviewed (commented):
LGTM otherwise.

Thoughts on changing the PR title to:
`Speeding up GPU clustering using smarter download strategy and memory allocations`?

FabianSchuetze commented:
> LGTM otherwise.
> 
> Thoughts on changing the PR title to:
> `Speeding up GPU clustering using smarter download strategy and memory allocations`?

Thanks, @kunaltyagi, for suggesting a new title. This title summarizes the PR indeed better by now.

kunaltyagi commented on gpu/segmentation/include/pcl/gpu/segmentation/impl/gpu_extract_clusters.hpp:
```diff
@@ -40,9 +40,22 @@
 #define PCL_GPU_SEGMENTATION_IMPL_EXTRACT_CLUSTERS_H_
 
 #include <pcl/gpu/segmentation/gpu_extract_clusters.h>
+namespace pcl {
```
```suggestion

namespace pcl {
```

FabianSchuetze commented:
Thanks @kunaltyagi for guiding me through the process and approving the PR!

larshg commented:
There seems to be missing an include for CUDA, see https://dev.azure.com/PointCloudLibrary/pcl/_build/results?buildId=18774&view=logs&jobId=149f4e41-9368-548e-d6d4-ff21d16f0ace&j=149f4e41-9368-548e-d6d4-ff21d16f0ace&t=30234468-9402-53e3-1a03-9eb9da598fc6.

FabianSchuetze commented:
Thank you, @larshg, for your comment! This is interesting - I will try to understand the error and try to address it.

larshg commented:
The segmentation module, just haven't contained any CUDA code directly. Only referenced other libraries, Containers, Octree etc and use their GPU implementation.

FabianSchuetze commented:
> The segmentation module, just haven't contained any CUDA code directly. Only referenced other libraries, Containers, Octree etc and use their GPU implementation.

Thanks for your comment! For the reason you described, I was initially hesitant to add Cuda code to the segmentation module but thought of extending the `DeviceArray<T>` API. In particular, #4689 outlines an idea for how to expand the API. Apart from not referencing Cuda directly in the segmentation module, we could also copy data ranges from the device in general. Should we maybe try to take that route?

kunaltyagi commented:
I think enhancing DeviceArray API seems a better options (more power to users and benefits in other parts of the module as well)

FabianSchuetze commented:
Thanks for your comment, @kunaltyagi, and also to @larshg again for spotting the issue! I am working on it :-)

FabianSchuetze commented:
@kunaltyagi I rebased my local branch to avoid the conflict mentioned here. After rebasing, I usually force-pushed the changes to the remote. That was fine as I was the only one who worked on this branch. However, I saw that you also committed changes to this branch. Do you have an idea what I can do to push the local rebase to the remote?

larshg commented:
You can just add that empty line yourself and force push it 😄 

https://github.com/PointCloudLibrary/pcl/pull/4677/commits/7d8bef572a449294d18e01e35e2b4eaa991be9b4

FabianSchuetze commented:
Thank you so much, Lars, for helping out again! :-)

kunaltyagi commented:
Builds failing

FabianSchuetze commented:
Puha! I think we are getting closer to an end. Thanks to the expansion of the device array API, we can avoid the Cuda calls in the segmentation module spotted by Lars.

kunaltyagi reviewed (approved):
LGTM

larshg commented on gpu/segmentation/include/pcl/gpu/segmentation/gpu_extract_clusters.h:
```diff
@@ -131,6 +131,7 @@ namespace pcl
           */
         void extract (std::vector<pcl::PointIndices> &clusters);
 
+
```
This should be removed?

FabianSchuetze replied:
Yes, this is a good idea. I have done that.

larshg commented on gpu/segmentation/include/pcl/gpu/segmentation/impl/gpu_extract_clusters.hpp:
```diff
@@ -100,80 +115,59 @@ pcl::gpu::extractEuclideanClusters (const typename pcl::PointCloud<PointT>::Ptr
       {
-        std::cout << " CPU: ";
+        PCL_DEBUG(" CPU: ");
+        //data.clear();
```
remove this commented code?

FabianSchuetze replied:
Oh, yes! This line shouldn't be there. Thank you!

larshg commented on gpu/segmentation/src/extract_clusters.cpp:
```diff
@@ -37,12 +37,26 @@
 #include <pcl/impl/instantiate.hpp>
 #include <pcl/point_types.h>
-//#include <pcl/gpu/segmentation/gpu_extract_labeled_clusters.h>
-#include <pcl/gpu/segmentation/impl/gpu_extract_clusters.hpp>
```
I think clang tidy would reorder these includes back, so might as well restore original order?

FabianSchuetze replied:
Hm, I can't restore the original lines then the code does not compile. I however did clang-format these lines. Your hypothesis about the order turned out to be correct.

larshg commented:
Apart from 3 really minor things, LGTM.

FabianSchuetze commented:
Thank you, Kunal and Lars, for the review! Lars, you are eagle-eyed! I did apply most of the changes you suggested - thank you.

larshg commented:
Just tested it. I went from something like 38 seconds, load of spam, faulty clusters, to 8 seconds, minimal spam, correct clusters, with my test pointcloud with about 120k points, thats segmented into 28 clusters with size of 3000-10.000 points.
However, the CPU version using Kd-tree, still does it in about 300 ms.

I noticed that line is still verbose:
https://github.com/PointCloudLibrary/pcl/blob/26756433993d7dbae4187b430a3b4024b05d2f54/gpu/segmentation/include/pcl/gpu/segmentation/impl/gpu_extract_clusters.hpp#L210

Could you fix this one as well 😄 ?

FabianSchuetze commented:
Thank you, Lars, for testing the program and for your feedback. Your message is bittersweet! I'm glad the results of the program itself were OK. Nevertheless, it would be wonderful to have a GPU version that performs faster than the CPU version. I will take a look at some GPU KDTree implementations for inspiration. Anyway - the noisy info got banished to the PCL_DEBUG macro, and it shan't be seen anymore during normal operation. Thanks for the feedback!

kunaltyagi commented on gpu/segmentation/include/pcl/gpu/segmentation/impl/gpu_extract_clusters.hpp:
```diff
@@ -214,7 +207,7 @@ pcl::gpu::EuclideanClusterExtraction<PointT>::extract (std::vector<pcl::PointInd
   // Extract the actual clusters
   extractEuclideanClusters<PointT> (host_cloud_, tree_, cluster_tolerance_, clusters, min_pts_per_cluster_, max_pts_per_cluster_);
-  std::cout << "INFO: end of extractEuclideanClusters " << std::endl;
+  PCL_DEBUG("INFO: end of extractEuclideanClusters ");
```
```suggestion
  PCL_DEBUG("INFO: end of extractEuclideanClusters\n");
```

kunaltyagi commented:
@mvieth Do you want to take a look? Or we can go ahead and squash-merge this

FabianSchuetze commented:
Thanks for approving, Lars and Kunal!

mvieth commented:
I had a quick look over the code, everything seems fine. And it definitely promises a great speedup, even if it might still not be able to compete with the pure CPU version. Thank you for working on this!
